lib/pbench/agent/tool_data_sink.py

```diff
@@ -1609,71 +1831,225 @@ def put_document(self, data_ctx, hostname):
+        else:
+            logger.exception("ERROR - failed to start the tool data sink")
+            ret_val = 9
+    except Exception:
+        logger.exception("ERROR - failed to start the tool data sink")
+        ret_val = 10
+    else:
+        ret_val = 0
+    return ret_val
+
+
+def daemon(
+    PROG,
+    redis_server,
+    redis_host,
+    redis_port,
+    pbench_bin,
+    pbench_run,
+    hostname,
+    tar_path,
+    cp_path,
+    param_key,
+    params,
+    optional_md,
+):
+    # Disconnect any existing connections to the Redis server.
+    redis_server.connection_pool.disconnect()
+    del redis_server
+
+    # Before we daemonize, flush any data written to stdout or stderr.
+    sys.stderr.flush()
+    sys.stdout.flush()
+
+    pidfile_name = f"{PROG}.pid"
+    pfctx = pidfile.PIDFile(pidfile_name)
+    with open(f"{PROG}.out", "w") as sofp, open(
+        f"{PROG}.err", "w"
+    ) as sefp, DaemonContext(
+        stdout=sofp,
+        stderr=sefp,
+        working_directory=os.getcwd(),
+        umask=0o022,
+        pidfile=pfctx,
+    ):
+        logger = get_logger(PROG, daemon=True)
+
+        # We have to re-open the connection to the redis server now that we
+        # are "daemonized".
+        logger.debug("re-constructing Redis server object")
+        try:
+            redis_server = redis.Redis(host=redis_host, port=redis_port, db=0)
+        except Exception as e:
+            logger.error(
+                "Unable to construct Redis server object, %s:%s: %s",
+                redis_host,
+                redis_port,
+                e,
+            )
+            return 7
+        else:
+            logger.debug("reconstructed Redis server object")
+        return driver(
+            PROG,
+            redis_server,
+            redis_host,
+            redis_port,
+            pbench_bin,
+            pbench_run,
+            hostname,
+            tar_path,
+            cp_path,
+            param_key,
+            params,
+            optional_md,
+            logger=logger,
+        )
+
+
+def main(argv):
+    _prog = Path(argv[0])
+    PROG = _prog.name
+    # The Tool Data Sink executable is in:
+    #   ${pbench_bin}/util-scripts/tool-meister/pbench-tool-data-sink
+    # So .parent at each level is:
+    #   _prog       ${pbench_bin}/util-scripts/tool-meister/pbench-tool-data-sink
+    #     .parent   ${pbench_bin}/util-scripts/tool-meister
+    #     .parent   ${pbench_bin}/util-scripts
+    #     .parent   ${pbench_bin}
+    pbench_bin = _prog.parent.parent.parent
+
     try:
         redis_host = argv[1]
         redis_port = argv[2]
         param_key = argv[3]
     except IndexError as e:
-        logger.error("Invalid arguments: %s", e)
+        print(f"{PROG}: Invalid arguments: {e}", file=sys.stderr)
         return 1
+    else:
+        if not redis_host or not redis_port or not param_key:
+            print(f"{PROG}: Invalid arguments: {argv!r}", file=sys.stderr)
+            return 1
```

Comment:
It saves one line by adding a test that is not needed.  See https://stackoverflow.com/questions/7604636/better-to-try-something-and-catch-the-exception-or-test-if-its-possible-first.

Reply:
It saves two lines, and it replaces with a simple `if` test all of the infrastructure required to capture the exception context (to wit, the `try` statement).  And, I think the result is better, more readable code, if for no other reason than it removes much of the indentation which requires the reader to track the scopes that the execution enters and leaves.

Reply:
OK, I'm impressed:  apparently Python has found some way to make `try` blocks really cheap (at least on X86_64) -- my suggested `if` construct is marginally _more_ expensive in the "normal" case.  (It's considerably faster in the exception case, but I don't think we're particularly interested in that, here.)

So, I guess this belongs in the "molehill" category rather than the "mountain" one.